### PR TITLE
A py script to combine multiple outputs of Evidence QC in one table

### PR DESCRIPTION
**Pull request by shadizaheri**
This is the python script created to add some of the metrics resulting from the EvidenceQC step to the out put data table.

### CODE REVIEW

**Comment by VJalili** [comment]
Thank you, @shadizaheri!!! It reads well, I made a few changes, mostly styling and abstracting methods, but please feel free to revert any if you prefer otherwise. Specifically, I made the following changes.

- There were multiple methods for reading the outputs of different methods (one for each of melt, manta, and wham). These methods were mostly implementing a similar functionality except for one array index. I refactored this into one method and made the array index as an input argument that can be adjusted w.r.t. to the method.

- You are composing some column names using filenames; e.g., from `bwa.Manta.QC.outlier.high` (filename), you compose `Manta_high` (column name). I think setting column names in the code w.r.t to the methods you call is easier and less error-prone than relying on the template of the filename. So, I made some changes similar to the following. 

```python
# Example 1
# from:
outlier_manta = filename_manta.split(".")[-4] + "_" + filename_manta.split(".")[-1]

# to
col_name = get_col_name("manta", outlier_type)

def get_col_name(caller: str, outlier_type: str) -> str:
    return f"{caller}_{outlier_type}_outlier"

# Example 2
# from
read_wham_outlier(filename_high_wham)
def read_wham_outlier(filename):
    outlier_wham = filename.split(".")[-4] + "_" + filename.split(".")[-1]
    ...
    outlier_wham_df.columns = ["#ID", str(outlier_wham) + "_outlier"]

# to
read_outlier(filename_high_wham, get_col_name("wham", "high"))
def read_outlier(filename, col_name):
    ...
    outlier_df.columns = ["#ID", col_name]
```

- I defined constant variable with its value set to `"#ID"`, and replaced all the references of `"#ID"` with the const var. This makes it less error-prone. 

- I fixed some issues with the command line args (optional vs. required, and the `-` to `_` conversion), and added a simple help message. Please feel free to extend the help messages if needed.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:109]: I am not sure how extracting sample ID from its URI is portable/reliable. But I'll leave it to you to decide how best the sample ID can be extracted.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:106]: You had set the ID to empty string, I have seen issues with that in particular when a line is trimmed before split by a delimiter. So, I changed the empty string with `NA`; however, I'm not sure what is the best approach here. Do you need this row?

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:193]: what is returned here? I guess it is `None`, if so, you may remove `return`.

**Review by shadizaheri — COMMENTED** [review]
Thank you so much, @VJalili  for your comments and suggestions for the script's style. Very helpful.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:193]: @VJalili This is supposed to return the final merged DataFrames in a "tsv" format.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:106]: @VJalili I think we need this row so that it can return `'0'` for the empty files (the files that we did not have any types of outliers).

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:109]: @VJalili using `os.path.split()` could also be an option, but it was not working here.

Comment (VJalili) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:109]: @shadizaheri do you have sample id in a different column of your data table? if so, that could be a better alternative, if not, it should be find to leave it as is.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:106]: Thanks! That makes sense to me.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:193]: Do you mean the filename?! I think since you're not using the returned value, it should be fine to remove `return`, but it should also be fine to leave it as is, it is up to you.

**Review by VJalili — APPROVED** [review]
Thank you, @shadizaheri!!! The script reads very well. I tested it and it runs as we discussed; though I would leave it to @mwalker174 or @epiercehoffman to check if it implements the logic we discussed.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:149]: When I did this I called `Counter()` on the concatenated sample ID columns from all the high outliers files, which seemed a little simpler/marginally more efficient because there were no dictionaries and you could count all of the callers at once

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:193]: Agreed, no need for a `return` since it is unused. Make sure to update the WDL with the expected output file name so that they match

Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:193]: Let's just make the prefix required so we can add a delimiter here
```suggestion
    return output_df.to_csv(f"{output_prefix}.evidence_qc_table.tsv", sep="\t", header=True, index=False)
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:200]:
```suggestion
        "estimated_copy_number_filename",
        help="Sets the filename containing estimated copy numbers per contig.")
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:203]:
```suggestion
        "median_cov_filename",
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:244]:
```suggestion
```

**Review by epiercehoffman — COMMENTED** [review]
Great, this is looking very close to ready! I think it will be really helpful. Thanks @shadizaheri for your work on this. I think overall the code looks organized, readable, well-documented, and 99% correct (though I haven't had a chance to do a test run).

There's only one real change to the actual output values that I've requested, which is related to the rounded chrX copy number used for batching. My other suggestions are for code style / simplicity / efficiency.

I'd also like to see an example of the output. Can you share an example output table and the input files used for the test? I'd also like to test the script as part of the EvidenceQc workflow because there is some preparation of the inputs involved - let me know what your plan is for finalizing the WDL and integrating the changes across branches.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:106]: @epiercehoffman It is therefore necessary to check if the file is blank. Empty files should also include outlier columns. Even empty files need IDs when merging based on the `'#ID'` key at the end.  I think it will be more clear once you run the code using the empty and non-empty files.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:109]: That makes sense, thanks!!

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:64]: This is an interesting perspective. I don't see a big difference between using 'sum' or 'counts'. I think having the code as it is (check below) can make further possible modifications easier.

```
    df_non_diploid = df_non_diploid.T.iloc[0:, 3:]
    new_c = df_non_diploid.iloc[0:, 0:]
    nondiploid_counts = new_c[new_c < 0.05].count()
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:244]: @epiercehoffman I'm not sure if I understand the suggestion here. Can you please explain the suggestion, here?

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: Thank you @VJalili and @epiercehoffman ! This is a valid point. The only reason that I had the `'return'` here, was to test the script by outputting a tsv file. Once we test the script and have the wdl written, we can remove the `'return'` line from here.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:244]: I think the prefix should be a required argument, so it does not need a default value

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:64]: I don't really care about sum vs. counts, but I think it's best to do this without making unnecessary copies of the dataframe. Why do you need new_c if it copies all the rows and columns?

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: The return has no impact on whether the TSV file is written. You can delete the `return` keyword and keep the rest of the line and it will write the TSV file - for testing & for use in the workflow.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:106]: Do you end up with an extra row for `NA` in the case of an empty file due to the outer join later? I see an extra row with no ID in your example output, which we would want to remove

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:48]: Can you name the output column `wgd_score`? Currently it is named `score` in the original WGD file, but that is ambiguous in the unified output table

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:66]: Let's name this column `nondiploid_bins` to indicate where it came from
```suggestion
    nondiploid_counts_df = pd.DataFrame(nondiploid_counts, columns=["nondiploid_bins"]).rename_axis(
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:199]: Dashes are best practice for command-line arguments, rather than underscores. When you access these options later as variables the dashes will be automatically converted to underscores. \

I think since there are so many arguments they should be named rather than positional. You can add a short argument name if you want. 

Some of the files are marked optional in the WDL task - make sure that is handled on the script level.
```suggestion
        "--estimated-copy-number-filename",
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:64]: I used `new_c`  before printing out/testing the code. We can easily delete `new_c` and modify as:

`    nondiploid_counts = df_non_diploid[df_non_diploid < 0.05].count()
`

Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:64]: Sounds good, let's go with that

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (VJalili) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:109]: Agree. I am more inclined to use [`.stem`](https://stackoverflow.com/a/47496703/947889), which should handle cases where you have multiple `.` in the filename, but `.split` should also be fine.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:199]: @epiercehoffman , this is a valid point, thanks for sharing. I also think that since these are required arguments, not optional, it might be better/easier if we go with underscore here. You can read more [here](https://stackoverflow.com/questions/12834785/having-options-in-argparse-with-a-dash)

Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:199]: You can make them required but still use `--`. I think that for this quantity of arguments they shouldn't all be positional.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: I see! I updated the script with ID_COL and checked to be sure that it's consistent all over the script.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: @epiercehoffman , I think there could be positional for the python script, at the same time I agree that it's better to not have them as positional! Thanks for the comment. I applied your comment in [this](https://github.com/broadinstitute/gatk-sv/commit/0086235f678fb29d32c88aa9fdd15ba7a76a71b8) commit.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: aha, now I see what you mean. Thanks for the comment and explanation. I removed `return`

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: Renamed to `wgd_score`.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: thanks for the comment! This got modified [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=outlier_type%3A%20low%20or%20high)

**Review by shadizaheri — COMMENTED** [review]
thanks for the comment! This got modified [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=outlier_type%3A%20low%20or%20high)

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:188]: It is less efficient to read the file again from disk. Using the outputs of the previous function would decrease the number of times that you read from disk as well as the number of dataframe operations. You directly call and use read_outlier within read_all_outlier, so it would be very simple to make those dataframes inputs to read_all_outlier instead of calling the function again.

Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:188]: We are in some sense using the output of other functions (`read_outlier`) inside the `read_all_outlier`. I don't think that making this change is necessary at this stage as I don't see much of a difference here.

Comment (epiercehoffman) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:188]: Let's touch base offline, I think it will be easier to explain!

Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py:188]: Sure! Meantime I thank you for your precise comment and conversation!
I think, for now, this is not going to cause any problems for us as I have checked the Pandas DataFrame memory usage for a different number of rows, columns, and data types. Even for the number of rows (in our case number of samples) in the order of 5e6 and the number of columns in the order of 40, we are not going to pass a couple of megabytes. 

So, I won't expect this to cause a problem for us, at the same time I definitely consider your comment for the future.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:246]: -w is used twice as a short argument name, here and for wham high outlier
```suggestion
        "-q", "--wham-qc-outlier-low-filename",
```

Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: very nice catch! I fixed it [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=%22%2Dr%22%2C%20%22%2D%2Dwham%2Dqc%2Doutlier%2Dlow%2Dfilename%22%2C).

**Review by shadizaheri — COMMENTED** [review]
Thanks for your work on this!

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (shadizaheri) [reply src/sv-pipeline/scripts/make_evidence_qc_table.py]: The files have been set as optional.

**Comment by shadizaheri** [comment]
@epiercehoffman and @VJalili thank you both for your very great comments and help on this PR! 
I believe I have addressed all the comments here.
You can test the script by running 
`python make_evidence_qc_table.py -y estimated_copy_numbers.txt -d bwr_medianCov.transposed.bed -g bwr_WGD_scores.txt -b binwise_CNV_qValues.bed -z bwr.Manta.QC.outlier.high -e bwr.Melt.QC.outlier.high -w bwr.Wham.QC.outlier.high -a bwr.Manta.QC.outlier.low -s bwr.Melt.QC.outlier.low -r bwr.Wham.QC.outlier.low -m melt_insert_size.tsv -o testing` command line. I am also attaching a zip file containing the input files as and a sample of the output file to this comment.
Please note that `melt_insert_size.tsv` is a tsv file that I exported from the data table to test the python script. Because the information for 'melt_insert_size.tsv' is already included in the data table as an output from the GatherSampleEvidence step, most likely we will not need this once we write the wdl as we can read this info from the data table in Terra.
[table_EvidenceQC_test.zip](https://github.com/broadinstitute/gatk-sv/files/10143461/table_EvidenceQC_test.zip)

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:240]: Not necessary to merge.
Why: I think we should update every reference to this file in this script to call it "mean" insert size instead of "Melt". We currently calculate the insert size during MELT, but as we work towards deprecating MELT, that will likely change, so we should be more flexible/accurate with the naming here.
Also: "melt" > "mean" elsewhere in script
```suggestion
        "-m", "--mean-insert-size-filename",
```

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:216]: Not necessary to merge: Scramble outlier files are missing here, so we should make sure to come back and update this script to handle them later. 
In the mean time, please remove the scramble inputs in the WDL task since the script does not use them.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:183]: Not necessary to merge / question of preference. When I tested without providing one of the outlier files, I noticed that it produced a column of NA values. I think I might prefer to omit the column if the corresponding file is not provided, but would be open to other views.

File: src/sv-pipeline/scripts/make_evidence_qc_table.py
Comment (epiercehoffman) [inline src/sv-pipeline/scripts/make_evidence_qc_table.py:117]: Not necessary to merge / question of preference. I think my preference would be to fill cells with 0, rather than `np.nan`, if the sample does not appear as an outlier at all in a specified category. This would allow the outlier columns to be stored as ints and may make downstream computation a bit simpler because users won't have to account for NaN values. But, I recognize that it's easier to have NaN values due to the dataframe merge later on, and I'm open to other views.

**Review by epiercehoffman — APPROVED** [review]
Thanks @shadizaheri for all the updates you've made this week and to both @shadizaheri and @VJalili for all your work on this! I think this is in good shape and can be merged - I've left a few comments of varying degrees of importance but I'm ok with merging as-is and returning to those later (more details below). I've tested it locally and it appears to work as expected. I'll include the details of my testing below for transparency.

Those are the important bits, the rest is details.

I've made a few comments, mainly related to required inputs/outputs. I have marked them as either "Address soon" if I believe they should ideally be addressed before merging the script (but ok if soon after) or "Not necessary to merge" if I think it's fine to come back to them in the future (without urgency). 

I approved the PR as-is because I think that when run as part of EvidenceQC in the way we typically expect (ie. with all of the inputs provided, except optionally omitting MELT outlier files), the script will perform as desired. So, the "Address soon" reflects that the script does not fully handle required/optional outputs correctly, but given the desire to get this PR in quickly, I would approve a merge as-is as long as we come back and fix these items soon to provide more flexibility with inputs. (Some comments relate to the WDL task as well so it would be good to read them over for those details even if we merge as-is.)

Testing details:
- Ran with all input files and checked: number of rows in output, no empty sample IDs, all values in expected range for metric, outlier counts add up across rows, rounded chrX as expected based on chrX CN (spot-checked), column names are good
- Spot-checked a couple of columns against the input files to make sure nothing got scrambled
- Ran with a gzipped estimated CN file and checked that the output is identical, since the provided test files were not gzipped but in the WDL several of the input files will be gzipped
- Ran removing one command line argument each time and checked for expected behavior. Worked fine to omit MELT outlier files, but I've left a few comments about optional/required inputs in other cases

**Pull request merged**
